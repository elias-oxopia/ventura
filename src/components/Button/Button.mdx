---
name: Button
menu: Components
route: /button
---

import { Playground, Props } from 'docz'
import { Activity, Trash2 } from '../../index.ts'
import Button from './Button.tsx'

# Button

Buttons make common actions more obvious and help users more easily perform them. Buttons use labels and sometimes icons to communicate the action that will occur when the user touches them.

### Best practices

- Group buttons logically into sets based on usage and importance.
- Ensure that button actions are clear and consistent.
- The main action of a group set can be a primary button.
- Select a single button variation and do not mix them.

## Basic usage

<Playground>
  <Button>Click me</Button>
</Playground>

## Button types

<Playground>
  <Button type="primary">Click me</Button>
  <br/>
  <Button type="secondary">Click me</Button>
</Playground>

## Button disabled

<Playground>
  <Button isDisabled>Click me</Button>
</Playground>

## Button loading

<Playground>
  <Button isLoading>Click me</Button>
</Playground>

## Button with icon

<Playground>
  <Button suffixIcon={<Activity />}>Click me</Button>
  <br/>
  <Button prefixIcon={<Trash2 />}>Click me</Button>
</Playground>

## API

<Props of={Button} />
