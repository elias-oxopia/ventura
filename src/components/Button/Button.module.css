.button {
  border-radius: var(--border-radius-small);
  line-height: 14px;
  font-size: 14px;
  padding: 10px 14px;
  border: none;
  cursor: pointer;
  display: flex;
}

.smallButton {
  padding: 3.5px 10px;
}

.typePrimary:hover {
  background-color: var(--color-primary-hover);
}

.typeSecondary:hover {
  background-color: var(--color-secondary-hover);
}

.typePrimary:focus {
  outline: none;
  box-shadow: 0 0 0 2px var(--color-focus);
}

.typeSecondary:focus {
  outline: none;
  box-shadow: 0 0 0 2px var(--color-focus);
}

.typePrimary,
.typePrimary:active {
  background-color: var(--color-primary);
}

.typeSecondary,
.typeSecondary:active {
  background-color: var(--color-secondary);
}

.button:disabled {
  color: var(--color-disabled-font);
  background-color: var(--color-disabled);
  cursor: default;
}

.label {
  line-height: inherit;
  max-height: 14px;
}

.smallLabel {
  font-size: calc(var(--font-size) - 2px);
}

.labelWithPrefixIcon {
  margin-left: 8px;
}

.labelWithSuffixIcon {
  margin-right: 8px;
}

.spinner {
  width: 14px;
  height: 14px;
  vertical-align: -2px;
}
