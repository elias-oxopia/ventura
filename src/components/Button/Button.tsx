import React, { MouseEventHandler } from 'react';
import classNames from 'classnames';

import Spinner from '../utils/Spinner/Spinner';

import cssReset from '../../css-reset.module.css';
import styles from './Button.module.css';

interface Props {
  children: React.ReactNode;
  type?: 'primary' | 'secondary';
  onClick?: MouseEventHandler<HTMLButtonElement>;
  className?: string;
  isDisabled?: boolean;
  isLoading?: boolean;
  isSubmit?: boolean;
  prefixIcon?: React.ReactElement;
  suffixIcon?: React.ReactElement;
  name?: string;
}

const Button: React.FC<Props> = ({
  children,
  type = 'primary',
  onClick,
  className,
  isDisabled,
  isLoading,
  isSubmit,
  prefixIcon,
  suffixIcon,
  name,
}: Props) => {
  const buttonClassNames = classNames(
    cssReset.ventura,
    styles.button,
    {
      [styles.typePrimary]: type === 'primary',
      [styles.typeSecondary]: type === 'secondary',
    },
    className,
  );

  const labelClassNames = classNames({
    [styles.labelWithPrefixIcon]: Boolean(prefixIcon) || isLoading,
    [styles.labelWithSuffixIcon]: Boolean(suffixIcon),
  });

  return (
    /* eslint-disable-next-line react/button-has-type */
    <button
      disabled={isDisabled || isLoading}
      className={buttonClassNames}
      type={isSubmit ? 'submit' : 'button'}
      onClick={isLoading || suffixIcon ? undefined : onClick}
      data-testid={`button-${name}`}
    >
      {isLoading ? <Spinner className={styles.spinner} /> : prefixIcon}
      <span className={labelClassNames}>{children}</span>
      {isLoading || suffixIcon}
    </button>
  );
};

export default Button;
