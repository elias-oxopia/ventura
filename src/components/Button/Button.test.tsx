import React from 'react';
import { fireEvent, render } from '@testing-library/react';

import { GitHub } from '../../index';
import Button from './Button';

describe('Button', () => {
  test('default snapshot', () => {
    const component = <Button>Click me</Button>;
    const { asFragment } = render(component);
    expect(asFragment()).toMatchSnapshot();
  });

  test('secondary button', () => {
    const component = <Button type="secondary">Click me</Button>;
    const { asFragment } = render(component);
    expect(asFragment()).toMatchSnapshot();
  });

  test('className prop', () => {
    const className = 'center';
    const component = (
      <Button className={className} name="class">
        Hey
      </Button>
    );
    const { getByTestId } = render(component);
    const buttonElement = getByTestId('button-class');

    const renderedClassNames = buttonElement.className.split(' ');
    expect(renderedClassNames).toContain(className);
    // className in prop should be the last in the row
    expect(renderedClassNames.indexOf(className)).toBe(renderedClassNames.length - 1);
  });

  test('onClick should be triggered when clicked', () => {
    const onClickFn = jest.fn();
    const component = (
      <Button onClick={onClickFn} name="click">
        Click me
      </Button>
    );
    const { getByTestId } = render(component);

    const button = getByTestId('button-click');
    fireEvent.click(button);

    expect(onClickFn).toHaveBeenCalledTimes(1);
  });

  test('disabled button', () => {
    const onClickFn = jest.fn();
    const component = (
      <Button isDisabled name="disabled">
        Click me
      </Button>
    );
    const { getByTestId } = render(component);
    const button = getByTestId('button-disabled');

    expect(button.hasAttribute('disabled')).toBe(true);
    expect(button.getAttribute('disabled')).not.toBe(false);

    fireEvent.click(button);
    expect(onClickFn).toHaveBeenCalledTimes(0);
  });

  test('loading button', () => {
    const onClickFn = jest.fn();
    const component = (
      <Button isLoading name="loading">
        Click me
      </Button>
    );
    const { asFragment, getByTestId } = render(component);

    expect(asFragment()).toMatchSnapshot();

    const button = getByTestId('button-loading');

    expect(button.hasAttribute('disabled')).toBe(true);
    expect(button.getAttribute('disabled')).not.toBe(false);

    fireEvent.click(button);
    expect(onClickFn).toHaveBeenCalledTimes(0);
  });

  test('button should have type="submit"', () => {
    const component = (
      <Button isSubmit name="submit">
        Click me
      </Button>
    );
    const { asFragment, getByTestId } = render(component);

    expect(asFragment()).toMatchSnapshot();

    const button = getByTestId('button-submit');

    expect(button.hasAttribute('type')).toBe(true);
    expect(button.getAttribute('type')).toBe('submit');
  });

  test('prefix icon button', () => {
    const component = <Button prefixIcon={<GitHub />}>Click me</Button>;
    const { asFragment } = render(component);
    expect(asFragment()).toMatchSnapshot();
  });

  test('suffix icon button', () => {
    const component = <Button suffixIcon={<GitHub />}>Click me</Button>;
    const { asFragment } = render(component);
    expect(asFragment()).toMatchSnapshot();
  });
});
