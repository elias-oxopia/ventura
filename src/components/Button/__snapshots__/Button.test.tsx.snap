// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button default snapshot 1`] = `
<DocumentFragment>
  <button
    class="button typePrimary"
    data-testid="button-undefined"
    type="button"
  >
    <span
      class=""
    >
      Click me
    </span>
  </button>
</DocumentFragment>
`;

exports[`Button loading button 1`] = `
<DocumentFragment>
  <button
    class="button typePrimary"
    data-testid="button-loading"
    disabled=""
    type="button"
  >
    <svg
      class="svg spinner"
      viewBox="0 0 150 150"
      x="0px"
      y="0px"
    >
      <circle
        class="circle"
        cx="75"
        cy="75"
        r="60"
      />
    </svg>
    <span
      class="labelWithPrefixIcon"
    >
      Click me
    </span>
  </button>
</DocumentFragment>
`;

exports[`Button prefix icon button 1`] = `
<DocumentFragment>
  <button
    class="button typePrimary"
    data-testid="button-undefined"
    type="button"
  >
    <svg
      fill="none"
      height="24"
      stroke="currentColor"
      stroke-linecap="round"
      stroke-linejoin="round"
      stroke-width="2"
      viewBox="0 0 24 24"
      width="24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M9 19c-5 1.5-5-2.5-7-3m14 6v-3.87a3.37 3.37 0 0 0-.94-2.61c3.14-.35 6.44-1.54 6.44-7A5.44 5.44 0 0 0 20 4.77 5.07 5.07 0 0 0 19.91 1S18.73.65 16 2.48a13.38 13.38 0 0 0-7 0C6.27.65 5.09 1 5.09 1A5.07 5.07 0 0 0 5 4.77a5.44 5.44 0 0 0-1.5 3.78c0 5.42 3.3 6.61 6.44 7A3.37 3.37 0 0 0 9 18.13V22"
      />
    </svg>
    <span
      class="labelWithPrefixIcon"
    >
      Click me
    </span>
  </button>
</DocumentFragment>
`;

exports[`Button secondary button 1`] = `
<DocumentFragment>
  <button
    class="button typeSecondary"
    data-testid="button-undefined"
    type="button"
  >
    <span
      class=""
    >
      Click me
    </span>
  </button>
</DocumentFragment>
`;

exports[`Button suffix icon button 1`] = `
<DocumentFragment>
  <button
    class="button typePrimary"
    data-testid="button-undefined"
    type="button"
  >
    <span
      class="labelWithSuffixIcon"
    >
      Click me
    </span>
    <svg
      fill="none"
      height="24"
      stroke="currentColor"
      stroke-linecap="round"
      stroke-linejoin="round"
      stroke-width="2"
      viewBox="0 0 24 24"
      width="24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M9 19c-5 1.5-5-2.5-7-3m14 6v-3.87a3.37 3.37 0 0 0-.94-2.61c3.14-.35 6.44-1.54 6.44-7A5.44 5.44 0 0 0 20 4.77 5.07 5.07 0 0 0 19.91 1S18.73.65 16 2.48a13.38 13.38 0 0 0-7 0C6.27.65 5.09 1 5.09 1A5.07 5.07 0 0 0 5 4.77a5.44 5.44 0 0 0-1.5 3.78c0 5.42 3.3 6.61 6.44 7A3.37 3.37 0 0 0 9 18.13V22"
      />
    </svg>
  </button>
</DocumentFragment>
`;
